pyfar/signal.py

```diff
@@ -444,6 +193,416 @@ def __setitem__(self, key, value):
             raise TypeError(
                     "Index must be int, not {}".format(type(key).__name__))
 
+
+class TimeData(Audio):
+    """Class for time data.
+
+    Objects of this class contain time data which is not directly convertable
+    to thefrequency domain, i.e., non-equidistantly spaced samples.
+
+    """
+    def __init__(self, data, times, comment=None, dtype=np.double):
+        """Init TimeData with data, and times.
+
+        Attributes
+        ----------
+        data : ndarray, double
+            Raw data in the time domain. The memory layout of data is 'C'.
+            E.g. data of shape (3, 2, 1024) has 3 x 2 channels with
+            1024 samples each.
+        times : ndarray, double
+            Times of the data in seconds. The number of times must match
+            the size of the last dimension of data.
+        comment : str
+            A comment related to the data. The default is None.
+        dtype : string, optional
+            Raw data type of the signal. The default is float64
+
+        """
+
+        Audio.__init__(self, 'time', comment, dtype)
+
+        # init data and meta data
+        self._data = np.atleast_2d(np.asarray(data, dtype=dtype))
+
+        self._n_samples = self._data.shape[-1]
+
+        self._times = np.atleast_1d(np.asarray(times).flatten())
+        if self._times.size != self.n_samples:
+            raise ValueError(
+                "The length of times must be data.shape[-1]")
+        if np.any(np.diff(self._times) <= 0) and len(times) > 1:
+            raise ValueError("Times must be monotonously increasing.")
+
+    @property
+    def time(self):
+        """The data in the time domain."""
+        if isinstance(self, Signal):
+            # the setter also converts the data from 'freq' to 'time'
+            self.domain = 'time'
+        else:
+            # this only changes the class internal variable
+            self._domain = 'time'
+        return self._data
+
+    @time.setter
+    def time(self, value):
+        data = np.atleast_2d(np.asarray(value))
+        self._domain = 'time'
+        self._data = data
+        self._n_samples = data.shape[-1]
+
+    @property
+    def n_samples(self):
+        """Number of samples."""
+        return self._n_samples
+
+    @property
+    def signal_length(self):
+        """The length of the signal in seconds."""
+        return self.times[-1]
+
+    @property
+    def times(self):
```

Comment:
Would a setter for `times` make sense, e.g. to shift in time?

Reply:
We might want it here (and also a setter for frequencies in `FrequencyData`). The Problem is that we do not want it in `Signal`, which inherits all methods of `TimeData` (https://medium.com/@george.shuklin/why-you-cant-remove-parent-methods-in-python-d4935e6d54a4).

I thus tend not to implement the setter - we should discuss this during the weekly.